### ISSUE TITLE
Default value for an enum causes CoercingSerializeException during introspection query

### ISSUE THREAD

**Issue by mjagus** [graphql-java/graphql-java#921]
Hi,

I'm trying to execute an introspection query against a certain schema containing a mutation with an optional enum input and a default value for it.

The result of introspection query does not include default value and following exception appears in application log:

```
graphql.schema.CoercingSerializeException: Invalid input for Enum 'ThreadSort'. Unknown value 'NEWEST_FIRST'
	at graphql.schema.GraphQLEnumType.getNameByValue(GraphQLEnumType.java:110)
	at graphql.schema.GraphQLEnumType.access$000(GraphQLEnumType.java:27)
	at graphql.schema.GraphQLEnumType$1.serialize(GraphQLEnumType.java:37)
	at graphql.language.AstValueHelper.serialize(AstValueHelper.java:174)
	at graphql.language.AstValueHelper.astFromValue(AstValueHelper.java:81)
	at graphql.introspection.Introspection.print(Introspection.java:118)
	at graphql.introspection.Introspection.lambda$static$1(Introspection.java:108)
	at graphql.execution.ExecutionStrategy.fetchField(ExecutionStrategy.java:222)
	at graphql.execution.ExecutionStrategy.resolveField(ExecutionStrategy.java:167)
	at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:55)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:392)
	at graphql.execution.ExecutionStrategy.lambda$completeValueForList$7(ExecutionStrategy.java:626)
	at graphql.execution.Async.each(Async.java:47)
	at graphql.execution.ExecutionStrategy.completeValueForList(ExecutionStrategy.java:604)
	at graphql.execution.ExecutionStrategy.completeValueForList(ExecutionStrategy.java:566)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:358)
	at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:328)
	at graphql.execution.ExecutionStrategy.lambda$resolveField$0(ExecutionStrategy.java:169)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at graphql.execution.ExecutionStrategy.resolveField(ExecutionStrategy.java:168)
	at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:55)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:392)
	at graphql.execution.ExecutionStrategy.lambda$completeValueForList$7(ExecutionStrategy.java:626)
	at graphql.execution.Async.each(Async.java:47)
	at graphql.execution.ExecutionStrategy.completeValueForList(ExecutionStrategy.java:604)
	at graphql.execution.ExecutionStrategy.completeValueForList(ExecutionStrategy.java:566)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:358)
	at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:328)
	at graphql.execution.ExecutionStrategy.lambda$resolveField$0(ExecutionStrategy.java:169)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at graphql.execution.ExecutionStrategy.resolveField(ExecutionStrategy.java:168)
	at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:55)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:392)
	at graphql.execution.ExecutionStrategy.lambda$completeValueForList$7(ExecutionStrategy.java:626)
	at graphql.execution.Async.each(Async.java:47)
	at graphql.execution.ExecutionStrategy.completeValueForList(ExecutionStrategy.java:604)
	at graphql.execution.ExecutionStrategy.completeValueForList(ExecutionStrategy.java:566)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:358)
	at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:328)
	at graphql.execution.ExecutionStrategy.lambda$resolveField$0(ExecutionStrategy.java:169)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at graphql.execution.ExecutionStrategy.resolveField(ExecutionStrategy.java:168)
	at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:55)
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:392)
	at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:328)
	at graphql.execution.ExecutionStrategy.lambda$resolveField$0(ExecutionStrategy.java:169)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at graphql.execution.ExecutionStrategy.resolveField(ExecutionStrategy.java:168)
	at graphql.execution.AsyncExecutionStrategy.execute(AsyncExecutionStrategy.java:55)
	at graphql.execution.Execution.executeOperation(Execution.java:153)
	at graphql.execution.Execution.execute(Execution.java:97)
	at graphql.GraphQL.execute(GraphQL.java:546)
	at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:488)
	at graphql.GraphQL.executeAsync(GraphQL.java:463)
	at graphql.GraphQL.execute(GraphQL.java:394)
	at graphql.servlet.GraphQLServlet.query(GraphQLServlet.java:338)
	at graphql.servlet.GraphQLServlet.doQuery(GraphQLServlet.java:300)
	at graphql.servlet.GraphQLServlet.doQuery(GraphQLServlet.java:296)
	at graphql.servlet.GraphQLServlet.lambda$new$1(GraphQLServlet.java:196)
	at graphql.servlet.GraphQLServlet.doRequest(GraphQLServlet.java:254)
	at graphql.servlet.GraphQLServlet.doPost(GraphQLServlet.java:272)
```

Running non-introspection queries seems to work as expected. Default value is actually respected.

Example query:
```
{
  __schema {
    queryType {
      fields {
        args {
          defaultValue
        }
      }
    }
  }
}
```

Example schema:
```
type Thread {
    id: ID!
    title: String!
    content: String!
}

enum ThreadSort {
    NEWEST_FIRST
    OLDEST_FIRST
    MOST_COMMENTS_FIRST
}

type Query {
    allThreads(sort: ThreadSort = NEWEST_FIRST) : [Thread!]!
}
```

**Comment by bbakerman** [comment]
Thanks for reporting this issue

I wrote this unit test to try and reproduce it.

```
class Issue921 extends Specification {

    def "can run introspection on a default value enum schema"() {
        def spec = '''
            type Thread {
                id: ID!
                title: String!
                content: String!
            }

            enum ThreadSort {
                NEWEST_FIRST
                OLDEST_FIRST
                MOST_COMMENTS_FIRST
            }
            
            type Query {
                allThreads(sort: ThreadSort = NEWEST_FIRST) : [Thread!]!
            }
            '''

        def qLSchema = TestUtil.schema(spec)
        def graphql = GraphQL.newGraphQL(qLSchema).build()

        when:
        def result = graphql.execute('''
                {
                  __schema {
                    queryType {
                      fields {
                        args {
                          defaultValue
                        }
                      }
                    }
                  }
                }   
            ''')

        then:
        result.errors.isEmpty()

        def json = JsonOutput.toJson(result.toSpecification())

        json == '{"data":{"__schema":{"queryType":{"fields":[{"args":[{"defaultValue":"NEWEST_FIRST"}]}]}}}}'
    }
}
```

But it works as expected.

Can you please check the assumptions here.

This was on latest master - can you confirm what version of graphql-java you are running please

I created this PR to demonstrate the bug report ...where it works

https://github.com/graphql-java/graphql-java/pull/926

**Comment by mjagus** [comment]
I did some digging and it seems that `SchemaParser` of graphql-java-tools creates `GraphQLEnumValueDefinition`s with enum objects as their values whereas `SchemaGenerator` of graphql-java simply creates them with String values. At least that's what happens in unit test environment.

During execution of an introspection query and when using graphql-java-tools, `GraphQLEnumType::getNameByValue` method will never find a proper match because an `equals` check happens between String and Enum object.

I have created PR for your PR to better demonstrate the test case: bbakerman/graphql-java#2 . The test should now fail with `CoercingSerializeException` as expected. I have used `NaturalEnumValuesProvider` to simulate behaviour of graphql-java-tools.
